# Basic Aggregations

When working with large datasets, your brain cannot process millions of individual rows. You need a way to summarize the data into single, meaningful numbers. This process is called **aggregation**.

Pandas provides built-in statistical functions to perform calculations across entire columns or rows.
*   **`.mean()`**: Calculates the average value of a column.
*   **`.sum()`**: Adds up all the values in a column.
*   **`.min()`** and **`.max()`**: Find the lowest and highest values.
*   **`.count()`**: Counts the number of non-null (valid) values in a column.
*   **`.describe()`**: Instantly generates a comprehensive statistical summary of all numeric columns.

These functions form the foundation of any analytical workflow.

###  Analogy: The Expense Box
Imagine you have a shoebox filled with paper store receipts.
*   Applying `.sum()` is like adding up the prices on every single receipt to find your **total spending**.
*   Applying `.mean()` is like finding the **average price** of an item you bought.
*   Applying `.min()` or `.max()` is like finding your **cheapest** or **most expensive** purchase.
*   Applying `.count()` is like counting **how many receipts** are in the box.

Instead of looking at each receipt one by one, these aggregations give you an instant, high-level understanding of your expenses.

### Code Examples

Let's use a simple dataset of software developers to see basic aggregations in action:


In [1]:
import pandas as pd

data = {
    'Name': ['Alex', 'John', 'Sarah', 'Emily', 'David'],
    'Language': ['TypeScript', 'Python', 'JavaScript', 'Python', 'TypeScript'],
    'Experience': [3, 5, 2, 8, 1],
    'Salary': [85000, 110000, 75000, 145000, 60000]
}

df = pd.DataFrame(data)
print(df)

    Name    Language  Experience  Salary
0   Alex  TypeScript           3   85000
1   John      Python           5  110000
2  Sarah  JavaScript           2   75000
3  Emily      Python           8  145000
4  David  TypeScript           1   60000


#### Aggregating an Entire DataFrame
If you call an aggregation function on a whole DataFrame, Pandas will try to calculate it for every column. However, some columns contain text (strings) which cannot be averaged. To avoid errors, you must use the parameter `numeric_only=True`:


In [2]:
# Calculate the mean of all numeric columns
print("--- Average Stats (Numeric Only) ---")
print(df.mean(numeric_only=True))

--- Average Stats (Numeric Only) ---
Experience        3.8
Salary        95000.0
dtype: float64


#### Aggregating a Specific Column (Preferred Way)
To find a statistic for just one column, select the column first and then call the aggregation method:

In [3]:
total_salary_payout = df['Salary'].sum()
average_experience = df['Experience'].mean()
max_salary = df['Salary'].max()

print(f"Total Salary: ${total_salary_payout}")
print(f"Average Experience: {average_experience} years")
print(f"Maximum Salary: ${max_salary}")

Total Salary: $475000
Average Experience: 3.8 years
Maximum Salary: $145000


#### Getting an Instant Overview with `.describe()`
Instead of writing individual lines of code for mean, min, max, and count, you can use `.describe()` to get them all at once:

In [4]:
print(df['Salary'].describe())

count         5.00000
mean      95000.00000
std       33354.16016
min       60000.00000
25%       75000.00000
50%       85000.00000
75%      110000.00000
max      145000.00000
Name: Salary, dtype: float64


### Common Pitfalls for Beginners
1.  **Trying to average text columns**: Writing `df.mean()` without setting `numeric_only=True` on a DataFrame that contains strings will raise a `TypeError` in newer versions of Pandas. Always remember to filter for numeric columns or specify the column directly.
2.  **Confusing `.count()` with `.size`**:
    *   `.count()` counts only **valid (non-null)** values. If a row has a `NaN` (missing) value, `.count()` ignores it.
    *   `df.shape[0]` or `.size` returns the total number of rows regardless of whether they have missing data.

#### Exercise 1 (Easy)
Given the following series of exam scores: `scores = pd.Series([85, 90, 78, None, 92, 88])`
1. Calculate the total sum of scores.
2. Find the average (mean) score. Note how Pandas handles the missing value (`None`/`NaN`) automatically!



In [5]:
import pandas as pd
scores = pd.Series([85, 90, 78, None, 92, 88])

print("Sum of scores:", scores.sum())
print("Average score:", scores.mean())

Sum of scores: 433.0
Average score: 86.6


*Explanation:* Pandas automatically ignores the missing value (`None`) when performing arithmetic aggregations like `.sum()` and `.mean()`. The mean is calculated by dividing 433 by 5 (the number of valid, non-null items).
